infrastructure/fw-trace/fw-trace.py

```diff
@@ -0,0 +1,407 @@
+			if ('targetTags' in item ) and (src_tag not in item['targetTags']):
+				if str(verbose) <> "None" : 
+					print ('  Egress firewall rule "%s" with target tags "%s" does not matches the source tag of "%s". Skipping to next rule' % (item['name'], item['targetTags'],src_tag))
+				continue
+
+
+
+			#search through the list of destination ranges to see if there is a match
+			#note that egress firewall rule destination filter of ranges an subnets are both stored as 'destinationRanges'
+			matched_range = False
+			for range in item['destinationRanges']:
+				if ipaddress.ip_address(dst) not in ipaddress.ip_network(range):
+					if str(verbose) <> "None" : 
+						print ('  Egress firewall rule "%s" with destination range of %s does not match the destination IP %s' % (item['name'],range, dst))
+				else:
+					if str(verbose) <> "None" : 
+						print ('  Egress firewall rule "%s" with destination range of %s matches the destination IP %s' % (item['name'],range, dst))
+					matched_range = True
+					break
+			if not matched_range:
+				if str(verbose) <> "None" : print ('  Egress firewall rule "%s" has no destination range that matches the destination IP %s. Skipping to next firewall rule' % (item['name'], dst))
+				continue
+				
+			#check to see if protocol/port match allow list
+			if 'allowed' in item:
+				allow_matched = False
+
+				for prot in item['allowed']:
+					if prot['IPProtocol'] == "all":
+						if str(verbose) <> "None" : print ('  Egress firewall rule "%s" allows all protocols and ports' % (item['name']))
+						allow_matched  = True
+						break
+				
+					matched_port = True
+					if ((prot['IPProtocol'] == "tcp") or (prot['IPProtocol'] == "udp")) and (prot['IPProtocol']  == protocol):
+						#checking to see if port matches any item is port list
+						matched_port = False
+						for p in prot['ports']:	
+							if p == port: 
+								matched_port = True
+								break
+							if '-' in p:				
+								if (int(port) >= int(p.partition('-')[0])) and (int(port) <= int(p.partition('-')[2])):
+									matched_port = True
+									break							
+														
+					if (protocol == prot['IPProtocol']) and matched_port :
+						if str(verbose) <> "None" : print ('  Egress firewall rule "%s" with allowed protocols %s matches the protocol %s:%s  ' % (item['name'], item['allowed'],protocol,port))
+						allow_matched  = True
+						break   
+
+			
+				if not allow_matched :
+					if str(verbose) <> "None" : print ('  Egress firewall rule "%s" does not allow or deny traffic on protocol %s:%s. Skipping to next firewall rule' % (item['name'], protocol, port))
+					continue
+				else:
+					print ('--->Egress firewall rule "%s" matched.  Traffic is allowed to egress to %s' % (item['name'],dst))
+					implicit_allow = False
+					break
+		
+			#check to see if protocol/port match denied list
+			if 'denied' in item:
+				denied_matched = False
```

Comment:
There seems to be a great deal of duplicated code here as well, can it be pulled out into other functions?